### ISSUE TITLE
slicearray init error when output array and input array are the same

### ISSUE THREAD

**Issue by maurocsound** [csound/csound#1039]
This works as expected:

```
instr 1
	kA1[]	fillarray	0, 1, 2, 3, 4, 5, 6, 7 ; <-- 8 elements
	kA2[]	slicearray	kA1, 2, 5 ; <-- 4 elements
		printk		0, lenarray:k(kA2)
		turnoff
endin
```
But if in the slicearray I use the same output array as the input array, Csound will stop the performance with "INIT ERROR: slice larger than original size". The following is the code that will exit with the init error:

```
instr 1
	kA1[]	fillarray	0, 1, 2, 3, 4, 5, 6, 7 ; <-- 8 elements
	kA1[]	slicearray	kA1, 2, 5 ; <-- 4 elements
		printk		0, lenarray:k(kA1)
		turnoff
endin
```
So, what happened here?

**Comment by jpffitch** [comment]
It adjusts the result array before any action.

⁣Sent from TypeApp ​

**Comment by maurocsound** [comment]
Ok, so, if it's not too difficult, I think that the slicearray opcode should manage this special case when the output and input array are the same, so that the above init error should be avoided.

**Comment by jpffitch** [comment]
You want this treatment for all array outputs?  I will think about it but it is a major coding change and I do not now how to do it consistently

**Comment by maurocsound** [comment]
I thought about it a lot, I looked at the source code (arrays.c) on github, and then I realized that the problem is that the statement '`kA1[] slicearray kA1, 2, 5`' works at init-time too. At init-time it changes the output array "virtual" size (with the 'tabensure' funtion in the c source code) and then, at k-time, it tries to change it again. Of course, it is this second attempt that causes the "INIT ERROR: slice larger than original size" error, being that at k-time the virtual size of kA1 is already changed.

For example, the following code doesn't give me errors, because I make Csound jump the slicearray at k-time: 

```
instr 1
	kA1[]	fillarray	0, 1, 2, 3, 4, 5, 6, 7 ; <-- 8 elements

		kgoto		jumpSlicearrayAtKtime
	kA1[]	slicearray	kA1, 2, 5
		jumpSlicearrayAtKtime:

		printk		0, lenarray:k(kA1)
		turnoff
endin
```

That's very confusing for the user, being that slicearray returns a k-array, so one would expect that the opcode doesn't change any k-rate variable at init-time.

So, in my opinion, to solve the problem, you should just avoid to make `k[] slicearray k[],i,i` work at init-time. Only the following should work at init-time: `i[] slicearray i[],i,i`.

Just out of curiosity, why do you make `k[] slicearray k[],i,i` work at init-time too? If the reason is because there is something in common to initialize for both the i-rate and k-rate versions, I think you should do that initialization in a separated c function. In this way, you could avoid to run the opcode at init-time even when the output is a k-variable, and that would solve the problem.

**Comment by jpffitch** [comment]
Simpler than that all that is needed to remove slicearray from running at 
i-time is the change 
{ "slicearray.k", sizeof(TABSLICE), 0, 3, "k[]", "k[]iip",
       (SUBR) tabslice, (SUBR) tabslice, NULL }
to
{ "slicearray.k", sizeof(TABSLICE), 0, 3, "k[]", "k[]iip",
       NULL, (SUBR) tabslice, NULL }

I am still unconvinced that this is the right thing to do as it was asked` 
for (by ?, Joachim?)

**Comment by jpffitch** [comment]
The iscussion should be o the csound-dev mailing list ot the main csound 
list.  Sugnificant users dio not read issues often.

**Comment by maurocsound** [comment]
Ok, so from the discussion so far in the Csnd-dev mailing list we found that on April, 9 2014 Joachim requested to make slicearray work with i-arrays too (before it worked only for k-arrays and, correctly, it ran only at k-time). After that John made some changes (April 15, 16 same year) to make slicearray work with i-arrays too, but since then, slicearray:k[] ran at i-time too, making impossible to use the same array for input and output (before these changes it was possible).

Summarizing, being that:
1. those changes removed (for mistake, I think) an existing and useful feature (the ability to use the same array for both the input and the output of the slicearray opcode);
2. from the reference manual, the regular behaviour for k-output opcodes should be to run and change their k-output only at k-time;
3. a k-rate opcode it's allowed to change its internal state at i-time for initializations purposes, but it should not change its k-output at i-time, unless it's "init" or some rare special case that "by definition" needs to return a k-value at i-time (see "genarray_i"); 
3. there is no single reason to make slicearray:k[] work at i-time too (correct me if I'm wrong).

So, for all the above reasons, I think it should be safe and correct to make slicearray:k[] run only at k-time. Of course, this should be exactly the same for "slicearray:S[]" too, being that, basically, a string is just another array.

**Comment by maurocsound** [comment]
Why the "wontfix" label?? 

Honestly I don't understand why you won't fix this. The only doubts could be on the behaviour of "slicearray:S[]", but I thought we all agree that "slicearray:k[]" should work only at k-time. 

I think the following schema is the correct behavior:

slicearray:i[]   ---> of course, should run only at i-time;
slicearray:k[]  ---> should run only at k-time (now it works at both times, confusing the user and making it impossible to use the same array for input and output);
slicearray:S[] ---> should run at both i-time and k-time because S-type doesn't have an implicit execution time (in fact, to avoid those doubts, the string opcodes, usually, have a double version, like "strcpy" and "strcpyk").

So, the only behaviour that should be changed is the "slicearray:k[]" one. The others should remain as they are now.

If someone has doubts on the above schema, can he/she share them with me, before labeling this as "wontfix" without explaining the reason? Thanks.

**Comment by jpffitch** [comment]
It is more complex ta you suggest.

Currently
  slicearray:i[] ---> runs at i-time;
  slicearray:k[] ---> runs only at k-time

i a not sure what slicearray:S[] does

The discussion was wider than that and gets into the behaviour of the 
internal function tabensure; chaging that created much worse problems

**Comment by maurocsound** [comment]
Ok, so if now slicearray:k[] works correctly only at k-time (why nobody told me that? :-D), it means that this specific issue has been solved and we can close it, right?

The label "wontfix" is inappropriate being that the problem I posed in the title of this thread has been solved...

About "slicearray:S[]" eventually we should create another issue and discuss it there. In this issue I was referring only at the k-time version of this opcode...

**Comment by maurocsound** [comment]
Mmmh... Forget it, there is still something wrong here... Now the slicearray:k[] works only at k-time and so I don't have anymore the INIT error, but if I run this code:

```
instr 1
	kA1[]	fillarray	0, 1, 2, 3, 4, 5, 6, 7 ; <-- 8 elements
		printf		"lenarray(kA1) before slicearray: %d\n", 1, lenarray:k(kA1)
	kA1	slicearray	kA1, 1, 4 ; <-- 4 elements
		printf		"lenarray(kA1) AFTER  slicearray: %d\n", 1, lenarray:k(kA1)
	
		printks		"kA1 after slicearray: { ", 0
	kCnt	=		0
	
	while (kCnt < lenarray:k(kA1)) do
			printf	"%d ", kCnt + 1, kA1[kCnt]
		kCnt	+=	1
	od
	
		printks		"}\n", 0
		turnoff
endin
```

The output is:
```
lenarray(kA1) before slicearray: 8
lenarray(kA1) AFTER  slicearray: 8
kA1 after slicearray: { 2 3 4 4 4 5 6 7 }
```

The user would expect:
```
lenarray(kA1) before slicearray: 8
lenarray(kA1) AFTER  slicearray: 4
kA1 after slicearray: { 1 2 3 4 4 5 6 7 }
```


So:
1) lenarray gives me the wrong array size after the slicearray: 8 instead of 4 (tabensure doesn't change anymore the array virtual size?).
2) it's like there is some trouble in the slicearray code in the indexing of the first elements?

**Comment by maurocsound** [comment]
Ok I found the problem for the (2) above.

The problem is that the array is changed two times... The first time in the init-pass, the second time in the k-pass, so the resulting array becomes strange:

i-pass: 0 1 2 3 4 5 6 7 --> 1 2 3 4 4 5 6 7
k-pass: 1 2 3 4 4 5 6 7 --> 2 3 4 4 4 5 6 7 (this is the resulting array from my code)

This happens because you make Csound run slicearray:k[] always 2 times, as you can see here:
https://github.com/csound/csound/blob/6d319136c69dac4127c721403e76c52b6bf7a354/Opcodes/arrays.c#L3879-L3880

I tried to remove the init-call to tabslice, recompiled Csound (finally I managed to build it on my Windows machine), and now my above code works (apart the lenarray problem):

 { "slicearray.k", sizeof(TABSLICE), 0, 2, "k[]", "k[]iip", 
**_NULL_**, (SUBR) tabslice, NULL }, 

What is for that init-call in tabslice? It appears to me that it's useless (and harmful in this case...).

**Comment by jpffitch** [comment]
Nonsense   In the change you suggest nothing is changed.  Te trea 2 means that there is NO init operation.

**Comment by maurocsound** [comment]
Mmmmh... So, maybe something went wrong in my building of Csound? I will try to download a version from appveyor and I'll check if it has the same behaviour. 

Meanwhile could you try this Csound program and tell me the output?

```
instr 1
	kA1[]	fillarray	0, 1, 2, 3, 4, 5, 6, 7 ; <-- 8 elements
		printf		"lenarray(kA1) before slicearray: %d\n", 1, lenarray:k(kA1)
	kA1	slicearray	kA1, 1, 4 ; <-- 4 elements
		printf		"lenarray(kA1) AFTER  slicearray: %d\n", 1, lenarray:k(kA1)
	
		printks		"kA1 after slicearray: { ", 0
	kCnt	=		0
	
	while (kCnt < lenarray:k(kA1)) do
			printf	"%d ", kCnt + 1, kA1[kCnt]
		kCnt	+=	1
	od
	
		printks		"}\n", 0
		turnoff
endin
```

**Comment by jpffitch** [comment]
SECTION 1:
new alloc for instr 1:
lenarray(kA1) before slicearray: 8
lenarray(kA1) AFTER  slicearray: 8
kA1 after slicearray: { 1 2 3 4 4 5 6 7 }
B  0.000 ..  0.100 T  0.100 TT  0.100 M:      0.0
Score finished in csoundPerform().
inactive allocs returned to freespace
end of score.              overall amps:      0.0
            overall samples out of range:        0
0 errors in performance
Elapsed time at end of performance: real: 0.005s, CPU: 0.005s

**Comment by maurocsound** [comment]
Ok, thanks. 
I found that at first I tried the above code in CsoundQT, but this used a different and older Csound lib  instead of the recent version I built from github. If I run the right Csound version from the command line, the output is like yours.

The lenarray question is another thing, it's related to tabensure and it's much more complex because it involves many other opcodes, like genarray. Eventually, it should be discussed in another issue.

So, if you agree, I think this particular "slicearray:k[]" problem is solved and then we can close this issue as "fixed".

Last question: why there is that first call in "(SUBR) tabslice, (SUBR) tabslice, NULL }," if with "2" it works only at k-time?
